crypto/bls12/bls12.go

```diff
@@ -155,231 +131,294 @@ func (agg AggregateSignature) Bitfield() crypto.Bitfield {
+	var aggregate bls12.PointG1
+	for _, pk := range publicKeys {
+		engine.G1.Add(&aggregate, &aggregate, pk.p)
+	}
+	return bls.coreVerify(&PublicKey{p: &aggregate}, message, signature, domain)
+}
 
-// CreateThresholdSignature creates a threshold signature from the given partial signatures.
-func (bc *bls12Crypto) CreateThresholdSignature(partialSignatures []consensus.Signature, _ consensus.Hash) (_ consensus.ThresholdSignature, err error) {
-	if len(partialSignatures) < bc.mods.Configuration().QuorumSize() {
-		return nil, crypto.ErrNotAQuorum
+// Sign creates a cryptographic signature of the given messsage.
+func (bls *bls12Base) Sign(message []byte) (signature consensus.QuorumSignature, err error) {
+	p, err := bls.coreSign(message, domain)
+	if err != nil {
+		return nil, fmt.Errorf("bls12: coreSign failed: %w", err)
 	}
-	sigs := make(map[hotstuff.ID]*Signature, len(partialSignatures))
-	for _, sig := range partialSignatures {
-		if _, ok := sigs[sig.Signer()]; ok {
-			err = multierr.Append(err, crypto.ErrPartialDuplicate)
-			continue
+	bf := crypto.Bitfield{}
+	bf.Add(bls.mods.ID())
+	return &AggregateSignature{sig: *p, participants: bf}, nil
+}
+
+// Combine combines multiple signatures into a single signature.
+func (bls *bls12Base) Combine(signatures ...consensus.QuorumSignature) (combined consensus.QuorumSignature, err error) {
+	if len(signatures) < 2 {
+		return nil, crypto.ErrCombineMultiple
+	}
+
+	g2 := bls12.NewG2()
+	agg := bls12.PointG2{}
+	var participants crypto.Bitfield
+	for _, sig1 := range signatures {
+		if sig2, ok := sig1.(*AggregateSignature); ok {
+			sig2.participants.RangeWhile(func(id hotstuff.ID) bool {
+				if participants.Contains(id) {
+					err = crypto.ErrCombineOverlap
+					return false
+				}
+				participants.Add(id)
+				return true
+			})
+			if err != nil {
+				return nil, err
+			}
+			g2.Add(&agg, &agg, &sig2.sig)
+		} else {
+			bls.mods.Logger().Panicf("cannot combine signature of incompatible type %T (expected %T)", sig1, sig2)
 		}
-		s, ok := sig.(*Signature)
+	}
+	return &AggregateSignature{sig: agg, participants: participants}, nil
+}
+
+// Verify verifies the given quorum signature against the message.
+func (bls *bls12Base) Verify(signature consensus.QuorumSignature, message []byte) bool {
+	s, ok := signature.(*AggregateSignature)
+	if !ok {
+		bls.mods.Logger().Panicf("cannot verify signature of incompatible type %T (expected %T)", signature, s)
+	}
+
+	l := s.Participants().Len()
+
+	if l == 1 {
+		id := firstParticipant(s.Participants())
+		pk, ok := bls.publicKey(id)
 		if !ok {
-			err = multierr.Append(err, fmt.Errorf("%w: %T", crypto.ErrWrongType, s))
-			continue
+			bls.mods.Logger().Warnf("Missing public key for ID %d", id)
+			return false
 		}
-		sigs[sig.Signer()] = s
+		return bls.coreVerify(pk, message, &s.sig, domain)
 	}
-	if len(sigs) < bc.mods.Configuration().QuorumSize() {
-		return nil, multierr.Combine(crypto.ErrNotAQuorum, err)
+
+	// else:
```

Comment:
Add more context; assume you referring to the `if l == 1`, meaning that `l > 1`.

Reply:
Yeah, will do. Here I would prefer to have an explicit else clause, because I think it makes the code clearer, but it goes against Go's style (and linters will complain).